File: 리트코드/0622-design-circular-queue/0622-design-circular-queue.java
public class Main {


}


class MyCircularQueue {

    int len;
    int[] queue;
    int front = 0;
    int rear = -1;
    public MyCircularQueue(int k) {
        queue = new int[k];
    }

    public boolean enQueue(int value) {
        if (isFull()) {
            return false;
        }
        rear = (rear + 1) % queue.length;
        queue[rear] = value;
        len++;
        return true;
    }

    public boolean deQueue() {
        if (isEmpty()) {
            return false;
        }
        front = (front + 1) % queue.length;
        len--;
        return true;
    }

    public int Front() {
        return isEmpty() ? -1 : queue[front];
    }

    public int Rear() {
        return isEmpty() ? -1 : queue[rear];
    }

    public boolean isEmpty() {
        return len == 0;
    }

    public boolean isFull() {
        return len == queue.length;
    }
}

/**
 * Your MyCircularQueue object will be instantiated and called as such:
 * MyCircularQueue obj = new MyCircularQueue(k);
 * boolean param_1 = obj.enQueue(value);
 * boolean param_2 = obj.deQueue();
 * int param_3 = obj.Front();
 * int param_4 = obj.Rear();
 * boolean param_5 = obj.isEmpty();
 * boolean param_6 = obj.isFull();
 */

File: 리트코드/0622-design-circular-queue/0622-design-circular-queue.kt
class MyCircularQueue(k: Int) {

    var len: Int = 0
    var front:Int = 0
    var rear : Int = -1
    var q : IntArray = IntArray(k)
    fun enQueue(value: Int): Boolean {
        return if(isFull()){
            false
        } else{
            rear = (rear+1) % q.size
            q[rear] = value
            len++
            true
        }
    }

    fun deQueue(): Boolean {
        return if(isEmpty()){
            false
        } else{
            front = (front+1) % q.size
            len--
            true
        }
    }

    fun Front(): Int {
        return if(isEmpty())
            -1
        else
            q[front]
    }

    fun Rear(): Int {
        return if(isEmpty())
            -1
        else
            q[rear]
    }

    fun isEmpty(): Boolean {
        return len == 0
    }

    fun isFull(): Boolean {
        return len == q.size
    }

}

/**
 * Your MyCircularQueue object will be instantiated and called as such:
 * var obj = MyCircularQueue(k)
 * var param_1 = obj.enQueue(value)
 * var param_2 = obj.deQueue()
 * var param_3 = obj.Front()
 * var param_4 = obj.Rear()
 * var param_5 = obj.isEmpty()
 * var param_6 = obj.isFull()
 */